src/impl/apple/keychain_helper.cpp

```diff
@@ -50,66 +51,91 @@ CFPtr<CFStringRef> convert_string(const std::string& string)
+    CFDictionaryAddValue(d.get(), kSecAttrAccount, account.get());
+    CFDictionaryAddValue(d.get(), kSecAttrService, service.get());
 #if TARGET_IPHONE_SIMULATOR
 #else
-    if (group) {
+    if (group)
         CFDictionaryAddValue(d.get(), kSecAttrAccessGroup, convert_string(*group).get());
-    }
 #endif
     return d;
 }
 
-std::vector<char> metadata_realm_encryption_key()
+/// Get the encryption key for a given service.
+util::Optional<std::vector<char>> try_getting_key(const CFStringPtr& account, const CFStringPtr& service)
 {
-    const size_t key_size = 64;
-    const std::string service = "io.realm.sync.keychain";
-    const std::string account = "metadata";
-
     auto search_dictionary = build_search_dictionary(account, service, none);
     CFDataRef retained_key_data;
     if (OSStatus status = SecItemCopyMatching(search_dictionary.get(), (CFTypeRef *)&retained_key_data)) {
-        if (status != errSecItemNotFound) {
+        if (status != errSecItemNotFound)
             throw KeychainAccessException(status);
-        }
 
-        // Key was not found. Generate a new key, store it, and return it.
-        std::vector<char> key(key_size);
-        arc4random_buf(key.data(), key_size);
-        auto key_data = adoptCF(CFDataCreate(nullptr, reinterpret_cast<const UInt8 *>(key.data()), key_size));
-        if (!key_data) {
-            throw std::bad_alloc();
-        }
-
-        CFDictionaryAddValue(search_dictionary.get(), kSecValueData, key_data.get());
-        if (OSStatus status = SecItemAdd(search_dictionary.get(), nullptr)) {
-            throw KeychainAccessException(status);
-        }
-
-        return key;
+        // Key was not found.
+        return none;
     }
-    CFPtr<CFDataRef> key_data = adoptCF(retained_key_data);
 
     // Key was previously stored. Extract it.
-    if (key_size != CFDataGetLength(key_data.get())) {
+    CFPtr<CFDataRef> key_data = adoptCF(retained_key_data);
+    if (key_size != CFDataGetLength(key_data.get()))
         throw std::runtime_error("Password stored in keychain was not expected size.");
-    }
 
     auto key_bytes = reinterpret_cast<const char *>(CFDataGetBytePtr(key_data.get()));
     return std::vector<char>(key_bytes, key_bytes + key_size);
 }
 
+void set_key(const std::vector<char>& key, const CFStringPtr& account, const CFStringPtr& service)
+{
+    auto search_dictionary = build_search_dictionary(account, service, none);
+    auto key_data = adoptCF(CFDataCreate(nullptr, reinterpret_cast<const UInt8 *>(key.data()), key_size));
+    if (!key_data)
+        throw std::bad_alloc();
+
+    CFDictionaryAddValue(search_dictionary.get(), kSecValueData, key_data.get());
+    if (OSStatus status = SecItemAdd(search_dictionary.get(), nullptr))
+        throw KeychainAccessException(status);
 }
+
+}   // anonymous namespace
+
+std::vector<char> metadata_realm_encryption_key(bool check_legacy_service)
+{
+    const CFStringPtr account = convert_string("metadata");
```

Comment:
These constants should use the `CFSTR` macro instead, so they end up in the `__cfstring` section of the binary rather than being allocated on the heap at runtime. There'd be no reason to use `CFPtr` with the resulting strings, either, as they'd have an eternal lifetime.